comictaggerlib/comiccacher.py

```diff
@@ -282,7 +290,7 @@ def add_volume_issues_info(self, volume_id: int, cv_volume_issues: list[CVIssues
                     "description": issue["description"],
                     "timestamp": timestamp,
                 }
-                self.upsert(cur, "issues", "id", issue["id"], data)
+                self.upsert(cur, "issues", "id", issue["id"], data) # Don't need source?
```

Comment:
upsert will not work as we need it it to, as it assumes that the value passed `issue["ID"]` will be a unique value, we currently do not have another source but that may change.

Reply:
```suggestion
                if 'source_name' in keys:
                  condition += ' AND source_name = ?'
                  vals.append(data['source_name'])
```
Adding that between:
```
        sql_ins = f"INSERT OR IGNORE INTO {tablename} ({keys})  VALUES ({ins_slots})"
        cur.execute(sql_ins, vals)

        sql_upd = f"UPDATE {tablename} SET {set_slots} WHERE {condition}"
        cur.execute(sql_upd, vals)
```
Is a fix but I think it'd be better to alter:
```
def upsert(self, cur: lite.Cursor, tablename: str, pkname: str, pkval: Any, data: dict[str, Any]) -> None:
```
to
```
def upsert(self, cur: lite.Cursor, tablename: str, condition: dict[str, Any], data: dict[str, Any]) -> None:
```
Condition: {'source_name': 'comicvine', 'id': 001}
Does that seems like a better idea?

Reply:
Thinking about this a bit more, the better solution might be to create the PK from the source name and issue/volume id: "cv_001" or similar. I think the DB will always be small enough that performance of a text PK shouldn't be an issue. Or, give each source a number and combine that.

I think this seems better all around?

Reply:
Altering the upsert function to take multiple keys is a good idea, creating a new primary key is even better (since we don't need to keep compatibility).
However sqlite lets us do this automatically using [insert or replace] and using multiple primary keys (I just looked this up)

[insert or replace]: https://www.sqlite.org/lang_replace.html

```patch
diff --git a/comictaggerlib/comiccacher.py b/comictaggerlib/comiccacher.py
index 7216279..6e1522d 100644
--- a/comictaggerlib/comiccacher.py
+++ b/comictaggerlib/comiccacher.py
@@ -90,28 +90,28 @@ class ComicCacher:
 
             cur.execute(
                 "CREATE TABLE Volumes("
-                + "id INT,"
+                + "id INT NOT NULL,"
                 + "name TEXT,"
                 + "publisher TEXT,"
                 + "count_of_issues INT,"
                 + "start_year INT,"
                 + "timestamp DATE DEFAULT (datetime('now','localtime')), "
-                + "source_name TEXT,"
-                + "PRIMARY KEY (id))"
+                + "source_name TEXT NOT NULL,"
+                + "PRIMARY KEY (id, source_name))"
             )
 
             cur.execute(
                 "CREATE TABLE AltCovers("
-                + "issue_id INT,"
+                + "issue_id INT NOT NULL,"
                 + "url_list TEXT,"
                 + "timestamp DATE DEFAULT (datetime('now','localtime')), "
-                + "source_name TEXT,"
-                + "PRIMARY KEY (issue_id))"
+                + "source_name TEXT NOT NULL,"
+                + "PRIMARY KEY (issue_id, source_name))"
             )
 
             cur.execute(
                 "CREATE TABLE Issues("
-                + "id INT,"
+                + "id INT NOT NULL,"
                 + "volume_id INT,"
                 + "name TEXT,"
                 + "issue_number TEXT,"
@@ -121,8 +121,8 @@ class ComicCacher:
                 + "site_detail_url TEXT,"
                 + "description TEXT,"
                 + "timestamp DATE DEFAULT (datetime('now','localtime')), "
-                + "source_name TEXT,"
-                + "PRIMARY KEY (id))"
+                + "source_name TEXT NOT NULL,"
+                + "PRIMARY KEY (id, source_name))"
             )
 
     def add_search_results(self, source_name: str, search_term: str, cv_search_results: list[CVVolumeResults]) -> None:
@@ -257,6 +257,7 @@ class ComicCacher:
                 pub_name = cv_volume_record["publisher"]["name"]
 
             data = {
+                "id": cv_volume_record["id"],
                 "source_name": source_name,
                 "name": cv_volume_record["name"],
                 "publisher": pub_name,
@@ -264,7 +265,7 @@ class ComicCacher:
                 "start_year": cv_volume_record["start_year"],
                 "timestamp": timestamp,
             }
-            self.upsert(cur, "volumes", "id", cv_volume_record["id"], data) # Don't need source?
+            self.upsert(cur, "volumes", data)
 
     def add_volume_issues_info(self, source_name: str, volume_id: int, cv_volume_issues: list[CVIssuesResults]) -> None:
 
@@ -279,6 +280,7 @@ class ComicCacher:
 
             for issue in cv_volume_issues:
                 data = {
+                    "id": issue["id"],
                     "volume_id": volume_id,
                     "source_name": source_name,
                     "name": issue["name"],
@@ -290,7 +292,7 @@ class ComicCacher:
                     "description": issue["description"],
                     "timestamp": timestamp,
                 }
-                self.upsert(cur, "issues", "id", issue["id"], data) # Don't need source?
+                self.upsert(cur, "issues", data)
 
     def get_volume_info(self, volume_id: int) -> CVVolumeResults | None:
 
@@ -382,13 +384,14 @@ class ComicCacher:
             timestamp = datetime.datetime.now()
 
             data = {
+                "id": issue_id,
                 "super_url": image_url,
                 "thumb_url": thumb_image_url,
                 "cover_date": cover_date,
                 "site_detail_url": site_detail_url,
                 "timestamp": timestamp,
             }
-            self.upsert(cur, "issues", "id", issue_id, data)
+            self.upsert(cur, "issues", data)
 
     def get_issue_select_details(self, issue_id: int) -> SelectDetails:
 
@@ -416,11 +419,10 @@ class ComicCacher:
 
             return details
 
-    def upsert(self, cur: lite.Cursor, tablename: str, pkname: str, pkval: Any, data: dict[str, Any]) -> None:
-        """This does an insert if the given PK doesn't exist, and an
-        update it if does
+    def upsert(self, cur: lite.Cursor, tablename: str, data: dict[str, Any]) -> None:
+        """This does an insert if the PK doesn't exist, and a
+        replace it if does
 
-        TODO: look into checking if UPDATE is needed
         TODO: should the cursor be created here, and not up the stack?
         """
 
@@ -443,13 +445,5 @@ class ComicCacher:
             ins_slots += "?"
             set_slots += key + " = ?"
 
-        keys += ", " + pkname
-        vals.append(pkval)
-        ins_slots += ", ?"
-        condition = pkname + " = ?"
-
-        sql_ins = f"INSERT OR IGNORE INTO {tablename} ({keys})  VALUES ({ins_slots})"
+        sql_ins = f"INSERT OR REPLACE INTO {tablename} ({keys}) VALUES ({ins_slots})"
         cur.execute(sql_ins, vals)
-
-        sql_upd = f"UPDATE {tablename} SET {set_slots} WHERE {condition}"
-        cur.execute(sql_upd, vals)
```